addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -183,24 +221,22 @@ def _timesheet_preprocess(self, vals):
             Overrride this to compute on the fly some field that can not be computed fields.
             :param values: dict values for `create`or `write`.
         """
-        # project implies analytic account
-        if vals.get('project_id') and not vals.get('account_id'):
-            project = self.env['project.project'].browse(vals.get('project_id'))
+        project = self.env['project.project'].browse(vals.get('project_id', False))
+        task = self.env['project.task'].browse(vals.get('task_id', False))
+        # task implies project
+        if task and not project:
+            project = task.project_id
+            if not project:
+                raise ValidationError(_('You cannot create a timesheet on a private task.'))
+            vals['project_id'] = project.id
+        if project and not vals.get('account_id'):
```

Comment:
Note: on 16.0, we use the `task_analytic_account_id`, but since not yet introduced in 14.0, I use the project.